portal-impl/src/com/liferay/portal/servlet/filters/aggregate/AggregateFilter.java

```diff
@@ -526,6 +528,49 @@ else if (minifierType.equals("js")) {
 			String finalContent = content;
 			String finalResourcePath = resourcePath;
 
+			Callable<String> callable = new Callable<String>() {
+
+				@Override
+				public String call() throws IOException {
+					String minifiedContent = null;
+
+					if (minifierType.equals("css")) {
+						minifiedContent = MinifierUtil.minifyCss(finalContent);
+					}
+					else {
+						minifiedContent = MinifierUtil.minifyJavaScript(
+							finalResourcePath, finalContent);
+					}
+
+					minifiedContent = StringBundler.concat(
+						_CSS_COMMENT_BEGIN,
+						URLUtil.getLastModifiedTime(resourceURL),
+						_CSS_COMMENT_END, StringPool.NEW_LINE, minifiedContent);
+
+					File tempFile = FileUtil.createTempFile();
+
+					FileUtil.write(tempFile, minifiedContent);
+
+					FileUtil.move(tempFile, cacheDataFile);
+
+					return minifiedContent;
+				}
+
+			};
+
+			PortalExecutorManager portalExecutorManager =
+				_portalExecutorManager;
+
+			if (portalExecutorManager == null) {
+				FutureTask<String> futureTask = new FutureTask<>(callable);
```

Comment:
@LisztLiszt You should just do `return callable.call()`. Additionally, please try to find a way to avoid concurrently executing this `Callable` because they may write the same file.

@yuhai Please help Hayden with the second part, thanks!